### PR TITLE
fix: open the file that match the activity log label, not preview.path

### LINKED ISSUE #322: Clicking a skill file in the activity log opens a different SKILL.md
### MonoCode version

0.1.51 (0.1.51)

### Operating system

macOS

### OS version

macOS 27

### Provider CLI

Codex

### What happened

Clicking the Zuse skill file in the agent’s activity log opens the project’s custom skill instead. The path shown in the activity log and the file opened in the preview pane do not match.

Environment
App: Monocode
OS: macOS
Model: GPT-6-Astra, High
Workspace: Git worktree
Monocode version: [add version]

Screenshots attached.

Image

Image

### Steps to reproduce

Steps to reproduce

1. Open a project containing .claude/skills/custom-skill/SKILL.md, with the Zuse skill available at ~/.codex/skills/zuse/SKILL.md.
2. Ask the agent to remove the project’s .zuse configuration.
3. The agent reads the Zuse skill. The activity log displays: /Users/<username>/.codex/skills/zuse/SKILL.md
4. Click that file entry.
5. Check the preview’s content and the file path displayed at the bottom.

Keep in mind, Zuse is just an example here but I imagine this happens to other skills too.

### PR DESCRIPTION
**Pull request by bluzername**
closes #322

## what

click a skill file in the activity log can open a different file than the path shown in the row. issue #322 has a good example: Zuse skill at `~/.codex/skills/zuse/SKILL.md` shown in the log, but click open the project own skill file instead.

## cause

in `ToolCallSummary` (`src/surfaces/AgentTranscript.tsx`), the row show `target`, which is parsed straight from the tool call label (for example `Read /Users/x/.codex/skills/zuse/SKILL.md`). but the click handler open `preview.path || target` instead of `target`. so when the tool metadata own `preview.path` field point somewhere else than the label say, the click open the wrong file with no warning. two files named `SKILL.md`, one under the project and one under a provider skills folder outside the project, is exactly the case where this show up.

## fix

always resolve the opened path from `target`, the same path already on screen, instead of `preview?.path || target`. what you click now always match what you read.

i also moved the label parsing (action/target/fileName/filePath) out of the component into a small pure function `resolveToolCallDisplay` in `src/surfaces/transcriptActivity.ts`. this repo vitest config is node-only (`src/**/*.test.ts`, no jsdom), so a component-level test was not an option here, moving the logic to a plain function is what make it testable at all.

## test

added `resolveToolCallDisplay` describe block in `transcriptActivity.test.ts`:
- a case with a label that carries a literal path different from `preview.path` (the bug from #322) - asserts `filePath` equal the label path, not `preview.path`.
- a case with no literal path in the label, to check it still fall back to `preview.path` correctly (no regression for the normal Read/Edit/Write rows).
- a case with a label that has no recognisable action.

confirmed red first: reverted just the one-line fix (kept `preview?.path || target`) and ran the new test, it fail with the exact mismatch from the issue. put the fix back, test pass.

ran `npm run check:web` after (`vitest run` + `tsc --noEmit`): 249 test files, 2724 tests, all pass, tsc clean. did not touch any Rust file so I did not run `check:rust`.

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved tool activity labels to show more accurate actions and targets.
  - File links now consistently open and display the resolved file path.
  - Paths shown in tool labels take precedence over conflicting preview paths.
  - Preview paths are used when labels do not include a file path.
  - Unrecognized tool actions no longer display misleading action or target details.
  - File previews now appear only when they match the resolved file.
  - Workspace paths beginning with `~` now resolve correctly for recognized home directories.

### CODE REVIEW

**Comment by hardbeat920** [comment]
Thanks for the fix, @bluzername. The approach makes sense. Before merging, could you handle these two edge cases?

- If the label and `preview.path` differ for a write, we can show one file’s diff but open another.
- Paths starting with `~/` currently resolve relative to the workspace instead of the home directory.

A couple of regression tests for these cases would be helpful too 🙏

**Comment by EricRasputin** [comment]
Thanks for helping with the fix btw. Can’t wait to get this merged as this happens quite a lot.

**Comment by bluzername** [comment]
Thanks for the review. I fixed both:

- The diff preview now only shows when the resolved preview path matches the file the row opens. If they differ, it falls back to the plain file control instead of showing the wrong diff.
- Paths that start with `~/` now expand to the home directory (when it can be recognized from cwd), instead of being joined onto the project path.

Added regression tests for both cases. All tests and typecheck pass.

**Comment by hardbeat920** [comment]
Hi @bluzername, thanks for working on this. I found a few cases that still need attention:

- Labels such as `Edit dependency versions` are treated as file paths, so the row does not fall back to `preview.path`.
- `previewMatchesFile` returns true when either path cannot be resolved, which can still show the wrong diff.
- `~/` paths fail when the project is outside a standard home directory. Could we use the existing `homeDir()` result instead of inferring it from cwd?

Thank you 🙏

**Comment by bluzername** [comment]
Hi @hardbeat920, thanks for the detailed cases. I fix all three:

- Label like `Edit dependency versions` is not a real filename, so now it fall back to the write preview's own path instead of trying to use the label text as a path. Test added for this.
- `previewMatchesFile` used to return true when either the preview path or the file path could not resolve at all, which is not the same as them agreeing. Now it only return true when both resolve and match, so an unresolved case falls back to the plain file control instead of a maybe-wrong diff. Test added.
- `~/` paths now prefer the real OS home directory. I add `setHomeDir` in `paths.ts` and prime it once at startup from the existing `fs.homeDir()` call (in `main.tsx`), so it works even when the project is outside a standard home shape. Falls back to the old cwd inference if that call has not resolved yet. Tests added for both cases (uses primed home dir, falls back when cleared).

Full `npx vitest run` (2734 tests) and `tsc --noEmit` both pass. Commit is 4ed9cd4.

**Comment by hardbeat920** [comment]
Hi @bluzername, thanks for the update. The previous issues are fixed. I found two remaining edge cases:

- homeDir() can finish after the app renders, and setHomeDir() does not trigger a rerender. This can leave ~/ rows unresolved.
- looksLikeFilePath() rejects valid labels such as
main.ts:12 and Windows paths ending in Dockerfile. 

Could we use resolveWorkspacePath() for this check?

The branch also still needs a rebase because #331 moved the affected files.

**Comment by hardbeat920** [comment]
Thanks for working on this @bluzername 

I think the original case can still open the wrong file, after the row passes an absolute skill path to onOpenFile, resolveFileOpenRequest can replace it with a same-named file from the project index. 
Could we preserve that path as exact?

Also, toolCallLabel uses preview.path to build the Read label before resolveToolCallDisplay runs. 

A test covering the full label-to-click path would help confirm the fix.

**Comment by bluzername** [comment]
Hi @hardbeat920, thanks for the two more cases. I fix both:

- homeDir() finishing after first render was the real bug. setHomeDir() only touch module state, it never make anything rerender. Now boot wait for that IPC call to settle before the first render (with a short timeout so a slow or hung call can not stall boot forever - if it does resolve after, setHomeDir() still run for anything rendered from then on).
- looksLikeFilePath() check the raw label text, so main.ts:12 lose its .ts once :12 is on the end, and a Windows Dockerfile path with backslashes never match the / check either. resolveWorkspacePath() already normalize both of these, so the trust check use that now instead.

I also pull the timeout-race logic into its own settleWithin() helper in concurrent.ts with its own tests, instead of writing it directly in main.tsx where it can not be tested. Added regression tests for both label cases too.

Full vitest run (2739 tests) and tsc --noEmit both pass. 3 unrelated tests timeout when the whole suite run together (FileTree, TabMiddleClick, ProjectNotificationSettings) but pass fine on their own, looks like sandbox timing not my change.

One more thing: the branch also needed a real rebase like you said, but I look at the diff against main and it is not just the file move from #331 - a lot of the src tree got reorganized (src/lib, src/chrome etc moved into src/app, src/features, src/shared, plus Rust and Cargo changes). That is bigger than I want to resolve blind in an unattended run, so I leave the branch as-is on top of the old base for now. Let me know if you want me to attempt the full merge, or if you would rather handle that side.

Commit is 7114b6d.

File: src/main.tsx [hunk 1]
```diff
@@ -12,11 +12,20 @@ import {
...
 import { consumeInstalledUpdate } from "./lib/updateNotice";
 import "./index.css";
 
 initAppearance();
 initSounds();
+// Prime the real home directory once so `~/` file references resolve exactly
+// rather than only being inferred from a session's cwd. Best-effort: a
+// project outside a recognisable home still falls back to that inference if
+// this IPC call is ever unavailable or slow.
+void homeDir()
```
[the remark this answers is not in the document]
Comment (bluzername) [reply src/main.tsx]: Fixed, boot now wait for the homeDir() IPC call to settle (with a short timeout so a slow/hung call can not block boot forever) before the first render, using a new settleWithin() helper. Commit 7114b6d.

**Comment by hardbeat920** [comment]
Thanks for the update!

The home-directory and label-path changes look helpful. 

I think the two issues in my last comment still need attention:
- An absolute skill path passed to onOpenFile can still be replaced with a same-named project file by resolveFileOpenRequest. Could we preserve that path as exact?
- toolCallLabel still builds the Read label from preview.path before resolveToolCallDisplay runs. Could we add a test covering the full path from the original tool label through the click?

The branch also still needs a rebase onto main before it can merge.

File: src/surfaces/transcriptActivity.ts [hunk 2]
```diff
@@ -201,6 +206,114 @@ export function editVerb(label: string): string {
...
+      : undefined;
+  const isFileVerb = (verb: string | undefined) =>
+    verb === "Read" || verb === "List" || verb === "Edit" || verb === "Write";
+  // A file-verb's captured target is only trustworthy as a path when it
+  // looks like one. Harnesses sometimes phrase these in plain English (e.g.
+  // "Edit dependency versions"), and treating that phrase itself as a
+  // filename both fails to resolve and shoulders out a real path the write
+  // preview already has.
+  const trustedLabelTarget =
+    labelTarget && (!isFileVerb(labelVerb) || looksLikeFilePath(labelTarget))
```
[the remark this answers is not in the document]
Comment (bluzername) [reply src/surfaces/transcriptActivity.ts]: Fixed, trustedLabelTarget now go through resolveWorkspacePath(labelTarget, cwd) instead of looksLikeFilePath(labelTarget) directly, so a location suffix like main.ts:12 and a Windows Dockerfile path both resolve correctly. Added regression tests for both. Commit 7114b6d.

File: src/lib/paths.ts [hunk 3]
```diff
@@ -89,6 +89,28 @@ export function joinPath(parent: string, relative: string): string {
...
+  const trimmed = trimSlash(cwd);
+  if (trimmed === "~") return undefined;
+  const parts = trimmed.split("/").filter(Boolean);
+  if (parts.length >= 2 && (parts[0] === "Users" || parts[0] === "home")) {
+    return `/${parts[0]}/${parts[1]}`;
+  }
+  if (
+    parts.length >= 3 &&
+    /^[A-Za-z]:$/.test(parts[0]) &&
+    parts[1] === "Users"
```
[the remark this answers is not in the document]
Comment (bluzername) [reply src/lib/paths.ts]: Fixed too, this one was not in @hardbeat920's last list but it is still real. parts[1] === "Users" is case-sensitive so a lowercase cwd like C:/users/dev/project never match. Compare with .toLowerCase() now, same as Windows itself treat it. Added a regression test. Commit 9935a65.

**Comment by bluzername** [comment]
One more small thing while I was in there: CodeRabbit's case-sensitive Windows Users match (from the round before your last message) was still open, not covered by the two cases you listed. Fixed that too, same pattern - .toLowerCase() compare, with a regression test. Commit 9935a65.

**Comment by EricRasputin** [comment]
Pls merge this soon. This issue is so annoying 😭
